### PR TITLE
Cache dashboard when navigating back to it via "Back button"

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/29500

Improve the "back to dashboard" flow by using the already loaded dashboard data to omit sending the BE request. Because the data stored in the store is normalized, we need to denormalize it before passing through. 

How to test:
- Open a dashboard with questions
- Click on a question title
- Without modifying the question, click the back button
- Your dashboard should appear without `/api/dashboard/:id` request.

### CODE REVIEW

File: frontend/src/metabase/dashboard/actions/data-fetching.js [hunk 1]
```diff
@@ -211,8 +231,10 @@ export const fetchDashboard = createThunkAction(
             },
           );
 
+      entities = entities ?? normalize(result, dashboard).entities;
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/data-fetching.js:234]: If we already got normalized `entities`, don't `normalize` the result again

File: frontend/src/metabase/dashboard/actions/data-fetching.js [hunk 2]
```diff
@@ -211,8 +231,10 @@ export const fetchDashboard = createThunkAction(
...
+
       return {
-        ...normalize(result, dashboard), // includes `result` and `entities`
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/data-fetching.js:215]: I haven't found where `result` was used, so only `entities` are passed now

**Review by cdeweyx — APPROVED** [review]

**Review by mazameli — APPROVED** [review]
This is dope

**Comment by WiNloSt** [comment]
@ranquild this isn't about this PR, but I'm curious if this epic https://github.com/metabase/metabase/issues/29500 eventually implements some kind of caching so that we can use the dashboard card result to display the question instead of having to call `/api/card/:card_id/query` again after visiting a question from clicking a dashboard card? I don't see that epic has any task regarding that, but it's in the product doc.

**Comment by ranquild** [comment]
@WiNloSt not currently. The shape of the data and permissions checks in these handlers are different. Maybe that's something for another milestone

**Comment by WiNloSt** [comment]
@ranquild I just realize this feature somehow breaks the dashboard slow timeout after going back from a question. I got this toast notification even when all dashboard cards are displayed.

Screenshot 2023-05-24 at 4 57 34 PM

The reason is that we have a logic that check `isRunning` then show that slow dashboard toast if `isRunning` is still true after 15 seconds.
https://github.com/metabase/metabase/blob/8ed81801f1fbd945dd94773e7edbf9a917a11867/frontend/src/metabase/dashboard/containers/DashboardApp/DashboardApp.jsx#L190-L193

But after going back to the dashboard, `isRunning` will stay `true`.
Screenshot 2023-05-24 at 4 58 14 PM

**Comment by ranquild** [comment]
@WiNloSt fixed, great catch

File: frontend/src/metabase/dashboard/hoc/DashboardData.jsx [hunk 3]
```diff
@@ -56,7 +56,9 @@ export default ComposedComponent =>
         initialize({ clear: !isNavigatingBackToDashboard });
 
         try {
-          await fetchDashboard(dashboardId, location && location.query);
+          await fetchDashboard(dashboardId, location && location.query, {
+            preserveDashboard: isNavigatingBackToDashboard,
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/hoc/DashboardData.jsx:60]: Based on my comment here (TODO: add a link to the comment in `frontend/src/metabase/dashboard/actions/data-fetching.js`), you can see how we're using `isNavigatingBackToDashboard` differently here, one without negation and one with negation, which could be mildly confusing.

**Review by WiNloSt — COMMENTED** [review]
Left some nitpicks, otherwise looks great 👍

**Review by WiNloSt — APPROVED** [review]

File: e2e/test/scenarios/dashboard/dashboard-drill.cy.spec.js [hunk 4]
```diff
@@ -1007,6 +1012,81 @@ describe("scenarios > dashboard > dashboard drill", () => {
...
+    cy.wait("@dashcardQuery");
+
+    getDashboardCard().within(() => {
+      cy.findByText("101.04").should("be.visible"); // table data
+      cy.findByText("Orders").click();
+      cy.wait("@cardQuery");
+    });
+
+    queryBuilderHeader().within(() => {
+      cy.findByDisplayValue("Orders").clear().type("Orders question").blur();
```
[the remark this answers is not in the document]
Comment (ranquild) [reply e2e/test/scenarios/dashboard/dashboard-drill.cy.spec.js:1061]: No, this was done to update the question and check that the change is reflected in the dashboard

File: frontend/src/metabase/dashboard/reducers.js [hunk 5]
```diff
@@ -266,7 +273,8 @@ const dashcardData = handleActions(
       next: (state, { payload: { cardId, dashcardId } }) =>
         assocIn(state, [dashcardId, cardId]),
     },
-    [Questions.actionTypes.UPDATE]: () => ({}),
+    [Questions.actionTypes.UPDATE]: (state, { payload: { object } }) =>
+      _.mapObject(state, data => dissoc(data, object.id)),
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/dashboard/reducers.js]: fixed

**Pull request merged**